core/src/apps/cardano/helpers/utils.py

```diff
@@ -64,9 +73,104 @@ def format_asset_fingerprint(policy_id: bytes, asset_name_bytes: bytes) -> str:
     return bech32.encode("asset", fingerprint)
 
 
+def format_script_hash(script_hash: bytes) -> str:
+    return bech32.encode(bech32.HRP_SCRIPT_HASH, script_hash)
+
+
+def format_key_hash(key_hash: bytes, is_shared_key: bool) -> str:
+    hrp = bech32.HRP_SHARED_KEY_HASH if is_shared_key else bech32.HRP_KEY_HASH
+    return bech32.encode(hrp, key_hash)
+
+
+def get_set_credential(
+    path: list[int],
+    key_hash: bytes | None,
+    script_hash: bytes | None,
+    pointer: CardanoBlockchainPointerType | None,
+) -> list[int] | bytes | CardanoBlockchainPointerType | None:
+    if path:
+        return path
+    elif key_hash:
+        return key_hash
+    elif script_hash:
+        return script_hash
+    elif pointer:
+        return pointer
+    else:
+        return None
+
+
+def get_set_credential_title(
+    path: list[int],
+    key_hash: bytes | None,
+    script_hash: bytes | None,
+    pointer: CardanoBlockchainPointerType | None,
+) -> str:
+    if path:
+        return "path"
+    elif key_hash:
+        return "key hash"
+    elif script_hash:
+        return "script"
+    elif pointer:
+        return "pointer"
+    else:
+        return ""
+
+
+def format_credential(
+    path: list[int],
+    key_hash: bytes | None,
+    script_hash: bytes | None,
+    pointer: CardanoBlockchainPointerType | None,
```

Comment:
I also have a vague idea about adding credential policy to the tuple, or an enum field indicating whether it is a payment or a staking credential and the policy will be determined later based on the value of the enum (I'm not sure if this makes sense).

Reply:
Hm, it's passed around only in utils and layout modules. I don't think it's necessary to add a class to simplify the passing around of these params for that. You simplify the passing of those parameters but make the calling of the function uglier because you have to create the `CredentialParams` object first. It would work, if `CardanoAddressParametersType` had two `credential` fields instead of the credential params being inlined, but that is IMHO not something we want to do now.

Reply:
On second thought - it might make sense to have a `CredentialParams` class and include the `format_credential` and `get_set_credential_title` functions in that. I'll get back to this.

Reply:
Looks good - https://github.com/vacuumlabs/trezor-firmware/pull/66#issuecomment-898288453

Reply:
Definitely much better.